File: api/gerar.js
/**
 * CNC Pro - API de Geração de G-code
 * 
 * Esta API recebe uma imagem, processa seus pixels e gera código G-code
 * para usinagem CNC. Suporta diferentes estratégias de usinagem:
 * - pocket: Escavação completa da área escura
 * - profile: Apenas o contorno da imagem
 * - contour: Segue o perímetro
 * - zigzag: Varredura simples
 * 
 * @author CNC Pro
 * @version 2.0
 */

// Importa a biblioteca Sharp para processamento de imagens
import sharp from 'sharp';

/**
 * Handler principal da função serverless Vercel
 * @param {Object} req - Requisição HTTP
 * @param {Object} res - Resposta HTTP
 */
export default async function handler(req, res) {
    // ========== CONFIGURAÇÃO CORS ==========
    // Permite que qualquer origem acesse a API (necessário para frontend separado)
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    
    // ========== TRATAMENTO DE PREFLIGHT (OPTIONS) ==========
    // Navegadores enviam OPTIONS antes do POST para verificar CORS
    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }
    
    // ========== VERIFICAÇÃO DO MÉTODO ==========
    // Apenas aceitamos requisições POST
    if (req.method !== 'POST') {
        return res.status(405).json({ 
            success: false, 
            error: 'Método não permitido. Use POST.' 
        });
    }
    
    try {
        // ========== EXTRAÇÃO DOS PARÂMETROS ==========
        const { 
            imagem,                    // Imagem em base64
            largura = 100,             // Largura da peça em mm
            altura = 100,              // Altura da peça em mm
            profundidade = 3,          // Profundidade total de corte em mm
            diametro = 3.175,          // Diâmetro da ferramenta em mm
            velocidade_corte = 800,    // Velocidade de corte em mm/min
            passo_corte = 0.5,         // Profundidade máxima por passe
            resolucao = 15,            // Passos por mm (qualidade)
            estrategia = 'pocket'      // Estratégia de usinagem
        } = req.body;
        
        // ========== VALIDAÇÃO DA IMAGEM ==========
        if (!imagem) {
            return res.status(400).json({ 
                success: false, 
                error: 'Nenhuma imagem fornecida' 
            });
        }
        
        // ========== PROCESSAMENTO DA IMAGEM ==========
        // Remove o cabeçalho base64 (ex: "data:image/png;base64,")
        const base64Data = imagem.split(',')[1];
        const imageBuffer = Buffer.from(base64Data, 'base64');
        
        // Calcula dimensões em pixels baseado na resolução
        const larguraPx = Math.max(1, Math.floor(largura * resolucao));
        const alturaPx = Math.max(1, Math.floor(altura * resolucao));
        
        // Processa a imagem com Sharp:
        // 1. Redimensiona para as dimensões de corte
        // 2. Converte para escala de cinza
        // 3. Extrai os pixels crus (0-255)
        const processedImage = await sharp(imageBuffer)
            .resize(larguraPx, alturaPx, { fit: 'fill' })
            .grayscale()
            .raw()
            .toBuffer();
        
        // Array de pixels (cada valor é brilho de 0=preto a 255=branco)
        const pixels = new Uint8Array(processedImage);
        
        // Calcula passo em mm por pixel
        const passoX = largura / larguraPx;
        const passoY = altura / alturaPx;
        
        // ========== GERAÇÃO DO G-CODE ==========
        const gcode = [];
        const dataHora = new Date().toLocaleString('pt-BR');
        
        // ---------- CABEÇALHO ----------
        gcode.push(`(CNC Pro - G-code Gerado Automaticamente)`);
        gcode.push(`(Data: ${dataHora})`);
        gcode.push(`(Dimensões: ${largura} x ${altura} mm)`);
        gcode.push(`(Profundidade: ${profundidade} mm)`);
        gcode.push(`(Ferramenta: ${diametro} mm)`);
        gcode.push(`(Resolução: ${resolucao} passos/mm)`);
        gcode.push(`(Estratégia: ${estrategia})`);
        gcode.push(`(Passo de corte: ${passo_corte} mm/passe)`);
        gcode.push(``);
        
        // Comandos iniciais do G-code
        gcode.push(`G90 ; Coordenadas absolutas`);
        gcode.push(`G21 ; Unidades em milímetros`);
        gcode.push(`M3 S10000 ; Liga o spindle`);
        gcode.push(`G0 Z5.000 ; Sobe para altura de segurança`);
        gcode.push(`G0 X0 Y0 ; Vai para origem`);
        gcode.push(``);
        
        // ========== CÁLCULO DOS PASSES DE PROFUNDIDADE ==========
        // Divide a profundidade total em passes para não sobrecarregar a ferramenta
        const numPasses = Math.ceil(profundidade / passo_corte);
        const incremento = profundidade / numPasses;
        
        let movimentosCorte = 0;
        let movimentosRapidos = 0;
        
        // ========== ESTRATÉGIA: ESCAVAÇÃO (POCKET) ==========
        // Ideal para: bolsos, áreas rebaixadas, gravação profunda
        // Preenche COMPLETAMENTE a área escura da imagem
        if (estrategia === 'pocket') {
            // Loop dos passes de profundidade
            for (let passe = 1; passe <= numPasses; passe++) {
                const zAtual = -(incremento * passe);
                gcode.push(`(Passe ${passe}/${numPasses} - Profundidade: ${zAtual.toFixed(3)} mm)`);
                gcode.push(`F${velocidade_corte} ; Velocidade de corte`);
                
                let emCorte = false;
                
                // Varredura em zig-zag para cada linha Y
                for (let y = 0; y < alturaPx; y++) {
                    const yPos = y * passoY;
                    
                    // Alterna direção a cada linha (otimiza movimentos)
                    if (y % 2 === 0) {
                        // Linha par: esquerda → direita
                        for (let x = 0; x < larguraPx; x++) {
                            const idx = y * larguraPx + x;
                            const brilho = pixels[idx];
                            const deveCortar = brilho < 128; // Escuro = cortar
                            
                            if (deveCortar) {
                                const xPos = x * passoX;
                                
                                if (!emCorte) {
                                    gcode.push(`G0 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    gcode.push(`G1 Z${zAtual.toFixed(3)}`);
                                    emCorte = true;
                                    movimentosRapidos++;
                                    movimentosCorte++;
                                } else {
                                    gcode.push(`G1 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    movimentosCorte++;
                                }
                            } else if (emCorte) {
                                gcode.push(`G0 Z5.000`);
                                emCorte = false;
                                movimentosRapidos++;
                            }
                        }
                    } else {
                        // Linha ímpar: direita → esquerda (zig-zag)
                        for (let x = larguraPx - 1; x >= 0; x--) {
                            const idx = y * larguraPx + x;
                            const brilho = pixels[idx];
                            const deveCortar = brilho < 128;
                            
                            if (deveCortar) {
                                const xPos = x * passoX;
                                
                                if (!emCorte) {
                                    gcode.push(`G0 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    gcode.push(`G1 Z${zAtual.toFixed(3)}`);
                                    emCorte = true;
                                    movimentosRapidos++;
                                    movimentosCorte++;
                                } else {
                                    gcode.push(`G1 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    movimentosCorte++;
                                }
                            } else if (emCorte) {
                                gcode.push(`G0 Z5.000`);
                                emCorte = false;
                                movimentosRapidos++;
                            }
                        }
                    }
                }
                
                // Garantir que a ferramenta está levantada no final
                if (emCorte) {
                    gcode.push(`G0 Z5.000`);
                    movimentosRapidos++;
                }
                
                gcode.push(``);
            }
        }
        
        // ========== ESTRATÉGIA: PERFIL (PROFILE) ==========
        // Ideal para: recortes, peças, silhuetas
        // Segue APENAS o contorno da imagem
        else if (estrategia === 'profile') {
            // Encontra todos os pixels de borda
            const bordas = [];
            for (let y = 0; y < alturaPx; y++) {
                for (let x = 0; x < larguraPx; x++) {
                    const idx = y * larguraPx + x;
                    const brilho = pixels[idx];
                    
                    if (brilho < 128) {
                        // Verifica se é borda (tem pelo menos um vizinho branco)
                        const vizinhos = [
                            y > 0 && pixels[(y-1) * larguraPx + x] >= 128,
                            y < alturaPx-1 && pixels[(y+1) * larguraPx + x] >= 128,
                            x > 0 && pixels[y * larguraPx + (x-1)] >= 128,
                            x < larguraPx-1 && pixels[y * larguraPx + (x+1)] >= 128
                        ];
                        
                        if (vizinhos.some(v => v)) {
                            bordas.push({ x: x * passoX, y: y * passoY });
                        }
                    }
                }
            }
            
            // Ordena bordas para caminho contínuo (algoritmo simples)
            const caminho = [];
            const visitados = new Set();
            let atual = bordas[0];
            
            while (atual && caminho.length < bordas.length) {
                caminho.push(atual);
                const idx = `${atual.x},${atual.y}`;
                visitados.add(idx);
                
                // Encontra o ponto mais próximo não visitado
                let proximo = null;
                let menorDistancia = Infinity;
                
                for (const ponto of bordas) {
                    const pontoIdx = `${ponto.x},${ponto.y}`;
                    if (!visitados.has(pontoIdx)) {
                        const dist = Math.hypot(ponto.x - atual.x, ponto.y - atual.y);
                        if (dist < menorDistancia) {
                            menorDistancia = dist;
                            proximo = ponto;
                        }
                    }
                }
                atual = proximo;
            }
            
            // Gera os passes de profundidade para o contorno
            for (let passe = 1; passe <= numPasses; passe++) {
                const zAtual = -(incremento * passe);
                gcode.push(`(Passe ${passe}/${numPasses} - Profundidade: ${zAtual.toFixed(3)} mm)`);
                gcode.push(`F${velocidade_corte} ; Velocidade de corte`);
                
                if (caminho.length > 0) {
                    gcode.push(`G1 X${caminho[0].x.toFixed(3)} Y${caminho[0].y.toFixed(3)} Z${zAtual.toFixed(3)}`);
                    movimentosCorte++;
                    
                    for (let i = 1; i < caminho.length; i++) {
                        gcode.push(`G1 X${caminho[i].x.toFixed(3)} Y${caminho[i].y.toFixed(3)}`);
                        movimentosCorte++;
                    }
                }
                
                gcode.push(`G0 Z5.000`);
                movimentosRapidos++;
                gcode.push(``);
            }
        }
        
        // ========== ESTRATÉGIA: ZIG-ZAG SIMPLES ==========
        // Ideal para: texturas, acabamentos, desbaste
        else {
            for (let passe = 1; passe <= numPasses; passe++) {
                const zAtual = -(incremento * passe);
                gcode.push(`(Passe ${passe}/${numPasses} - Profundidade: ${zAtual.toFixed(3)} mm)`);
                gcode.push(`F${velocidade_corte} ; Velocidade de corte`);
                
                let emCorte = false;
                
                for (let y = 0; y < alturaPx; y++) {
                    const yPos = y * passoY;
                    
                    if (y % 2 === 0) {
                        for (let x = 0; x < larguraPx; x++) {
                            const idx = y * larguraPx + x;
                            const deveCortar = pixels[idx] < 128;
                            
                            if (deveCortar) {
                                const xPos = x * passoX;
                                
                                if (!emCorte) {
                                    gcode.push(`G0 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    gcode.push(`G1 Z${zAtual.toFixed(3)}`);
                                    emCorte = true;
                                    movimentosRapidos++;
                                    movimentosCorte++;
                                } else {
                                    gcode.push(`G1 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    movimentosCorte++;
                                }
                            } else if (emCorte) {
                                gcode.push(`G0 Z5.000`);
                                emCorte = false;
                                movimentosRapidos++;
                            }
                        }
                    } else {
                        for (let x = larguraPx - 1; x >= 0; x--) {
                            const idx = y * larguraPx + x;
                            const deveCortar = pixels[idx] < 128;
                            
                            if (deveCortar) {
                                const xPos = x * passoX;
                                
                                if (!emCorte) {
                                    gcode.push(`G0 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    gcode.push(`G1 Z${zAtual.toFixed(3)}`);
                                    emCorte = true;
                                    movimentosRapidos++;
                                    movimentosCorte++;
                                } else {
                                    gcode.push(`G1 X${xPos.toFixed(3)} Y${yPos.toFixed(3)}`);
                                    movimentosCorte++;
                                }
                            } else if (emCorte) {
                                gcode.push(`G0 Z5.000`);
                                emCorte = false;
                                movimentosRapidos++;
                            }
                        }
                    }
                }
                
                if (emCorte) {
                    gcode.push(`G0 Z5.000`);
                    movimentosRapidos++;
                }
                
                gcode.push(``);
            }
        }
        
        // ========== RODAPÉ ==========
        gcode.push(`M5 ; Desliga o spindle`);
        gcode.push(`G0 Z5.000 ; Sobe para altura de segurança`);
        gcode.push(`G0 X0 Y0 ; Retorna à origem`);
        gcode.push(`M30 ; Fim do programa`);
        
        // ========== RESPOSTA DE SUCESSO ==========
        return res.status(200).json({
            success: true,
            gcode: gcode.join('\n'),
            linhas: gcode.length,
            movimentos_corte: movimentosCorte,
            movimentos_rapidos: movimentosRapidos,
            profundidade_max: profundidade,
            num_passes: numPasses,
            dimensoes_px: { largura: larguraPx, altura: alturaPx }
        });
        
    } catch (error) {
        // ========== TRATAMENTO DE ERROS ==========
        console.error('Erro na API:', error);
        return res.status(500).json({ 
            success: false, 
            error: error.message 
        });
    }
}

File: api/gerar.py
"""
API Serverless para Vercel
Endpoint de geração de G-code
"""

import base64
import json
import re
import sys
import os
from pathlib import Path
from io import BytesIO

# Adiciona o diretório raiz ao path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from PIL import Image
import numpy as np


class ConfigGerador:
    """Configuração do gerador"""
    def __init__(self, config_dict):
        self.passos_por_mm = config_dict.get('passos_por_mm', 10.0)
        self.profundidade_max_corte = config_dict.get('profundidade_max_corte', 0.5)
        self.altura_seguranca_z = config_dict.get('altura_seguranca_z', 5.0)
        self.velocidade_corte = config_dict.get('velocidade_corte', 800.0)
        self.velocidade_rapida = config_dict.get('velocidade_rapida', 3500.0)
        self.limiar_preto_branco = config_dict.get('limiar_preto_branco', 128)
        self.diametro_ferramenta = config_dict.get('diametro_ferramenta', 0.4)


class GeradorGCode:
    """Gerador de G-code para serverless"""
    
    def __init__(self, config):
        self.config = ConfigGerador(config)
        self.passo_mm = 1.0 / self.config.passos_por_mm
    
    def gerar(self, imagem_data, dimensoes_mm, profundidade_total):
        """Gera G-code a partir de dados da imagem"""
        try:
            # Carrega imagem do base64
            if ',' in imagem_data:
                imagem_data = imagem_data.split(',')[1]
            
            img_bytes = base64.b64decode(imagem_data)
            img = Image.open(BytesIO(img_bytes)).convert('L')
            
            largura_mm, altura_mm = dimensoes_mm
            
            # Redimensiona
            largura_px = max(1, int(largura_mm * self.config.passos_por_mm))
            altura_px = max(1, int(altura_mm * self.config.passos_por_mm))
            img = img.resize((largura_px, altura_px), Image.Resampling.LANCZOS)
            
            # Binariza
            pixels = np.array(img)
            binario = pixels < self.config.limiar_preto_branco
            
            # Calcula passes
            profundidade_total = -abs(profundidade_total)
            num_passes = max(1, int(np.ceil(abs(profundidade_total) / self.config.profundidade_max_corte)))
            incremento = abs(profundidade_total) / num_passes
            
            # Gera G-code
            gcode = self._gerar_cabecalho(dimensoes_mm, profundidade_total)
            
            for i in range(1, num_passes + 1):
                z = -incremento * i
                gcode.extend(self._gerar_passagem(binario, z, i, num_passes))
            
            gcode.extend(self._gerar_rodape())
            
            return gcode
            
        except Exception as e:
            raise Exception(f"Erro ao gerar G-code: {str(e)}")
    
    def _gerar_cabecalho(self, dimensoes, profundidade):
        from datetime import datetime
        largura, altura = dimensoes
        return [
            "(CNC Pro - G-code Gerado Online)",
            f"(Data: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')})",
            f"(Dimensões: {largura:.2f} x {altura:.2f} mm)",
            f"(Profundidade: {abs(profundidade):.2f} mm)",
            "",
            "G90 G21",
            f"M3 S10000",
            f"G0 Z{self.config.altura_seguranca_z:.3f} F{self.config.velocidade_rapida:.1f}",
            "G0 X0 Y0",
            ""
        ]
    
    def _gerar_passagem(self, imagem, z_atual, passe, total):
        altura, largura = imagem.shape
        gcode = [
            f"(Passe {passe}/{total} Z: {z_atual:.3f} mm)",
            f"F{self.config.velocidade_corte:.1f}"
        ]
        
        em_corte = False
        
        for y in range(altura):
            if y % 2 == 0:
                x_range = range(largura)
            else:
                x_range = range(largura - 1, -1, -1)
            
            for x in x_range:
                if imagem[y, x]:
                    x_mm = x * self.passo_mm
                    y_mm = y * self.passo_mm
                    
                    if not em_corte:
                        gcode.append(f"G0 X{x_mm:.3f} Y{y_mm:.3f}")
                        gcode.append(f"G1 Z{z_atual:.3f}")
                        em_corte = True
                    else:
                        gcode.append(f"G1 X{x_mm:.3f} Y{y_mm:.3f}")
                elif em_corte:
                    gcode.append(f"G0 Z{self.config.altura_seguranca_z:.3f}")
                    em_corte = False
        
        if em_corte:
            gcode.append(f"G0 Z{self.config.altura_seguranca_z:.3f}")
        
        return gcode
    
    def _gerar_rodape(self):
        return [
            "",
            "M5",
            f"G0 Z{self.config.altura_seguranca_z:.3f}",
            "G0 X0 Y0",
            "M30"
        ]


def handler(request, context):
    """
    Função principal do Vercel Serverless Function
    """
    # Configura CORS
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'POST, GET, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
    }
    
    # Resposta para preflight (OPTIONS)
    if request.method == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    # Apenas POST é permitido
    if request.method != 'POST':
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({'success': False, 'error': 'Método não permitido'})
        }
    
    try:
        # Parse do body
        body = json.loads(request.body)
        
        imagem = body.get('imagem')
        largura = float(body.get('largura', 100))
        altura = float(body.get('altura', 100))
        profundidade = float(body.get('profundidade', 3))
        resolucao = float(body.get('resolucao', 10))
        limiar = int(body.get('limiar', 128))
        
        if not imagem:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'success': False, 'error': 'Nenhuma imagem fornecida'})
            }
        
        # Configura o gerador
        config = {
            'passos_por_mm': resolucao,
            'profundidade_max_corte': 0.5,
            'altura_seguranca_z': 5.0,
            'velocidade_corte': 800,
            'velocidade_rapida': 3500,
            'limiar_preto_branco': limiar,
            'diametro_ferramenta': 0.4
        }
        
        gerador = GeradorGCode(config)
        
        # Gera o G-code
        gcode = gerador.gerar(imagem, (largura, altura), -profundidade)
        
        # Analisa estatísticas
        movimentos_corte = len([l for l in gcode if l.startswith('G1')])
        movimentos_rapidos = len([l for l in gcode if l.startswith('G0')])
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'success': True,
                'gcode': '\n'.join(gcode),
                'linhas': len(gcode),
                'movimentos_corte': movimentos_corte,
                'movimentos_rapidos': movimentos_rapidos
            })
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'success': False, 'error': str(e)})
        }